diff --git a/src/utils.ts b/src/utils.ts
index 369f6cec22bfcb39..d6145f43b2aeee6b 100644
--- a/src/utils.ts
+++ b/src/utils.ts
@@ -203,7 +203,7 @@ export class Utils {
         // Expand all variables
         evalStr = this.expandTextWith(evalStr, {
             unescape: JSON.stringify("$"),
-            variable: (name) => JSON.stringify(envs[name] ?? null),
+            variable: (name) => JSON.stringify(envs[name] ?? null).replaceAll("\\\\", "\\"),
         });
 
         // Convert =~ to match function
diff --git a/tests/rules.test.ts b/tests/rules.test.ts
index 5fed69935395ebd3..1f2bb903b08648de 100644
--- a/tests/rules.test.ts
+++ b/tests/rules.test.ts
@@ -238,6 +238,12 @@ test("Complex parentheses junctions regex fail - case insensitive", () => {
     expect(val).toBe(false);
 });
 
+test("Regex with escape characters from variable", () => {
+    const ruleIf = "$TAG =~ $TAG_REGEX";
+    const val = Utils.evaluateRuleIf(ruleIf, {TAG: "prefix/1.0.0", TAG_REGEX: "/^prefix\\/.+/"});
+    expect(val).toBe(true);
+});
+
 test("https://github.com/firecow/gitlab-ci-local/issues/350", () => {
     let rules, rulesResult, variables;
     rules = [
