diff --git a/src/parser-includes.ts b/src/parser-includes.ts
index 5161f76102343f20..30071c3488320cf6 100644
--- a/src/parser-includes.ts
+++ b/src/parser-includes.ts
@@ -72,7 +72,11 @@ export class ParserIncludes {
                 const files = await globby([value["local"].replace(/^\//, ""), ...excludedGlobs], {dot: true, cwd});
                 for (const localFile of files) {
                     const content = await Parser.loadYaml(`${cwd}/${localFile}`, {inputs: value.inputs || {}});
-                    excludedGlobs.push(`!${localFile}`);
+                    // If the file is included with `inputs:` set, the user might want to include it
+                    // again with different values, so we cannot exclude it.
+                    if(!value["inputs"]) {
+                        excludedGlobs.push(`!${localFile}`);
+                    }
                     includeDatas = includeDatas.concat(await this.init(content, depth, opts));
                 }
             } else if (value["project"]) {
diff --git a/tests/test-cases/include-inputs/input-templates/basic-example/.gitlab-ci.yml b/tests/test-cases/include-inputs/input-templates/basic-example/.gitlab-ci.yml
index 5127163ad98727d1..aa47fb3e160c5659 100644
--- a/tests/test-cases/include-inputs/input-templates/basic-example/.gitlab-ci.yml
+++ b/tests/test-cases/include-inputs/input-templates/basic-example/.gitlab-ci.yml
@@ -2,8 +2,12 @@
 include:
   - local: './.gitlab-ci-input-template.yml'
     inputs:
-      message: hello world
+      message: hello website
       options_input: website
+  - local: './.gitlab-ci-input-template.yml'
+    inputs:
+      message: hello db
+      options_input: db

 stages:
   - test
diff --git a/tests/test-cases/include-inputs/integration.include-inputs.test.ts b/tests/test-cases/include-inputs/integration.include-inputs.test.ts
index 5f03b4a87d3f626b..0d8ad2b6d0e83ae7 100644
--- a/tests/test-cases/include-inputs/integration.include-inputs.test.ts
+++ b/tests/test-cases/include-inputs/integration.include-inputs.test.ts
@@ -23,9 +23,14 @@ stages:
   - .post
 scan-website:
   script:
-    - echo hello world
-    - echo hello world
-    - echo hello world`;
+    - echo hello website
+    - echo hello website
+    - echo hello website
+scan-db:
+  script:
+    - echo hello db
+    - echo hello db
+    - echo hello db`;

     expect(writeStreams.stdoutLines[0]).toEqual(expected);
 });
